### PR TITLE
feat: refactor production test and improvs

### PR DESCRIPTION
**Pull request by FKobus**
TODO:

- splitted modules as own bundle
- multiple entries
- commons

### CODE REVIEW

File: src/production/ProductionContext.ts [hunk 1]
```diff
@@ -1,30 +1,51 @@
...
+}
+
+export function createProductionContext(ctx: Context, modules?: Array<IModule>): IProductionContext {
+  let entries;
+
+  /**
+   * @todo:
+   * Not sure about this check
+   */
+  if (!modules) {
```
Comment (FKobus) [inline src/production/ProductionContext.ts]: from tests we provide it with a modules array...
do you have an idea how to solve this more properly?

Comment (nchanged) [reply src/production/ProductionContext.ts]: we need to throw an exception... Probably in ModuleResolver, a fatal error that the files weren't resolved

File: src/production/phases/FinalPhase.ts [hunk 2]
```diff
@@ -1,20 +1,43 @@
...
+  await ctx.ict.resolve();
+
+  for (const module of modules) {
     if (module.isExecutable) {
-      module.transpile();
+      const transformerResult = module.transpile();
+
+      for (const item of transformerResult.requireStatementCollection) {
+        item.statement.callee.name = BUNDLE_RUNTIME_NAMES.ARG_REQUIRE_FUNCTION;
+        item.statement.arguments[0].value = module.moduleSourceRefs[item.statement.arguments[0].value].id;
```
Comment (nchanged) [inline src/production/phases/FinalPhase.ts]: so many things can go wrong here ;-)

Comment (FKobus) [reply src/production/phases/FinalPhase.ts]: i know.. will fix this (added a comment)

File: src/core/fusebox.ts [hunk 3]
```diff
@@ -1,57 +1,54 @@
...
         props.handler(new UserHandler(ctx));
       }
-      bundleProd(ctx);
 
-      // return bundleProd(ctx);
-      return Promise.resolve();
+      try {
+        await bundleProd(ctx);
+      } catch (e) {
+        ctx.log.error(`Fatal error: ${e}`);
```
Comment (nchanged) [inline src/core/fusebox.ts]: `ctx.fatal`

File: src/moduleResolver/ModuleResolver.ts [hunk 4]
```diff
@@ -80,22 +80,32 @@ function initModule(props: { absPath: string; bundleContext: IBundleContext; ctx
...
-          statement: source,
-        });
-
-        module.moduleSourceRefs[source] = resolvedModule;
-        // re-writing the reference
-        item.statement.callee.name = BUNDLE_RUNTIME_NAMES.ARG_REQUIRE_FUNCTION;
-        item.statement.arguments[0].value = resolvedModule.id;
-        delete item.statement.arguments[0].raw;
+        // no value.. import() or require();
+        if (!item.statement.arguments[0]) continue;
```
Comment (nchanged) [inline src/moduleResolver/ModuleResolver.ts]: we still need to make an `ARG_REQUIRE_FUNCTION` of it. And pass `-1` or `0` as an argument. (IMHO) so we could pick it up and runtime and throw an error

File: src/moduleResolver/ModuleResolver.ts [hunk 5]
```diff
@@ -80,22 +80,32 @@ function initModule(props: { absPath: string; bundleContext: IBundleContext; ctx
...
+            bundleContext,
+            ctx,
+            importType: item.importType,
+            parent: module,
+            statement: source,
+          });
+          if (!resolvedModule) continue;
+          module.moduleSourceRefs[source] = resolvedModule;
+          // re-writing the reference
+          item.statement.callee.name = BUNDLE_RUNTIME_NAMES.ARG_REQUIRE_FUNCTION;
```
Comment (nchanged) [inline src/moduleResolver/ModuleResolver.ts]: How about making this re-usable?

File: src/moduleResolver/ModuleResolver.ts [hunk 6]
```diff
@@ -80,22 +80,32 @@ function initModule(props: { absPath: string; bundleContext: IBundleContext; ctx
...
+          // we're importing some weird stuff... require(1) ie.
+          if (typeof source !== 'string') continue;
+          const resolvedModule = resolve({
+            bundleContext,
+            ctx,
+            importType: item.importType,
+            parent: module,
+            statement: source,
+          });
+          if (!resolvedModule) continue;
```
Comment (nchanged) [inline src/moduleResolver/ModuleResolver.ts]: Let's make sure that if we're not able to resolve that module we do:

* What I mentioned above
*  Also make sure that logging is in place

**Pull request merged**